### PR TITLE
Insights API version 1

### PR DESCRIPTION
**Pull request by SharmaAjay19**
This is version 1 of Insights API, which can run detectors, and surface any critical/warning insights.
The API can take various kind of inputs, like detector ids to run, a support topic id or a text verbatim (for which we get detectors using Project Shield).

    {
      "detectors": ["appDownAnalysis"],
      "supportTopicId": "32542218",
      "text": "app is down"
    }

The output is a list of insights with each insight having below attributes (the structure is basic and will keep evolving with time):

    {
              "status": 0,
              "title": "HTTP Server Errors Detected ",
              "description": "HTTP Servers errors indicate something wrong either with Application code or by the platform serving the requests.",
              "detailsLink": {
                  "type": 0,
                  "text": "HTTP Server Errors",
                  "uri": "https://applens.azurewebsites.net//subscriptions/52d78239-4706-40d2-b6aa-527492dc8eb1/resourceGroups/AccountServiceResourceGroup-centralus/providers/Microsoft.Web/sites/accountservice-prod-centralus/detectors/httpservererrors"
              },
              "table": null,
              "additionalDetails": null
    }

For details link, the API returns the detector link considering the resource uri as the base URL.

Currently this will be used with our CRI automation system to surface issues/solutions on incidents. Will keep making changes and improvements to this as we go further integrating it with various systems.

### CODE REVIEW

File: src/Diagnostics.RuntimeHost/Utilities/UriElements.cs [hunk 1]
```diff
@@ -45,6 +45,7 @@ public class UriElements
         public const string Configurations = "configurations";
         public const string KustoClusterMappings = Configurations + "/kustoClusterMappings";
         public const string UniqueResourceId = "/{id}";
+        public const string DiagnosticReport = "diagreport";
```
Comment (ShekharGupta1988) [inline src/Diagnostics.RuntimeHost/Utilities/UriElements.cs:48]: shall we just call it 'diagnostics'?

File: src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs [hunk 2]
```diff
@@ -450,6 +450,192 @@ private Exception FlattenIfAggregatedException(Exception ex)
...
+                        }
+                    }
+                }
+            }
+
+            var filteredDetectorsToRun = detectorsToRun.Union(allChildrenOfAnalyses).ToList();
+            DiagnosticsETWProvider.Instance.LogRuntimeHostInsightCorrelation(cxt.OperationContext.RequestId, "DiagnosticReportAPI.DetectorsRun", cxt.OperationContext.Resource.SubscriptionId,
+                    cxt.OperationContext.Resource.ResourceGroup, cxt.OperationContext.Resource.Name, correlationId, $"Running detectors: {JsonConvert.SerializeObject(filteredDetectorsToRun.Select(x => x.Metadata.Id).ToList())}");
+
+            if (filteredDetectorsToRun.Count>0)
```
Comment (ShekharGupta1988) [inline src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs:491]: nitpick - just format the whole document once :)

File: src/Diagnostics.RuntimeHost/Utilities/UriElements.cs (hunk 1 above)
Comment (SharmaAjay19) [reply src/Diagnostics.RuntimeHost/Utilities/UriElements.cs]: Done

File: src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs (hunk 2 above)
Comment (SharmaAjay19) [reply src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs]: Done

File: src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs [hunk 3]
```diff
@@ -450,6 +450,193 @@ private Exception FlattenIfAggregatedException(Exception ex)
...
+                }
+            }
+
+            var filteredDetectorsToRun = detectorsToRun.Union(allChildrenOfAnalyses).ToList();
+            DiagnosticsETWProvider.Instance.LogRuntimeHostInsightCorrelation(cxt.OperationContext.RequestId, "DiagnosticReportAPI.DetectorsRun", cxt.OperationContext.Resource.SubscriptionId,
+                    cxt.OperationContext.Resource.ResourceGroup, cxt.OperationContext.Resource.Name, correlationId, $"Running detectors: {JsonConvert.SerializeObject(filteredDetectorsToRun.Select(x => x.Metadata.Id).ToList())}");
+
+            if (filteredDetectorsToRun.Count > 0)
+            {
+                foreach (var detector in filteredDetectorsToRun)
```
Comment (ShekharGupta1988) [inline src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs:493]: Is it possible to run all detectors in parallel ?

**Review by ShekharGupta1988 — APPROVED** [review]

File: src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs (hunk 3 above)
Comment (SharmaAjay19) [reply src/Diagnostics.RuntimeHost/Controllers/DiagnosticControllerBase.cs]: Done

**Pull request merged**